eth/downloader/receiptreference.go

```diff
@@ -0,0 +1,125 @@
+func initReceiptReferences() {
+	// if already loaded, return
+	if len(userDepositNoncesReference) != 0 {
+		return
+	}
+	files, err := receiptReference.ReadDir(receiptReferencePath)
+	if err != nil {
+		log.Warn("Receipt Correction: Failed to read reference directory", "err", err)
+		return
+	}
+	for _, file := range files {
+		// load to map
+		fpath := path.Join(receiptReferencePath, file.Name())
+		bs, err := receiptReference.ReadFile(fpath)
+		if err != nil {
+			log.Warn("Receipt Correction: Failed to read reference data", "err", err)
+			continue
+		}
+		udns := userDepositNonces{}
+		err = gob.NewDecoder(bytes.NewReader(bs)).Decode(&udns)
+		if err != nil {
+			log.Warn("Receipt Correction: Failed to decode reference data", "err", err)
+			continue
+		}
+		userDepositNoncesReference[udns.ChainID] = udns
+	}
+}
+
+// correctReceipts corrects the deposit nonce in the receipts using the reference data
+// prior to Canyon Hard Fork, DepositNonces were not cryptographically verifiable.
+// As a consequence, the deposit nonces found during Snap Sync may be incorrect.
+// This function inspects transaction data for user deposits, and if it is found to be incorrect, it is corrected.
+// The data used to correct the deposit nonce is stored in the userDepositData directory,
+// and was generated with the receipt reference tool in the optimism monorepo.
+func correctReceipts(receipts types.Receipts, transactions types.Transactions, blockNumber uint64, chainID uint64) types.Receipts {
+	initReceiptReferences()
+	// get the reference data for the chain currently running
+	refs, ok := userDepositNoncesReference[chainID]
+	if !ok {
+		log.Trace("Receipt Correction: No data source for chain", "chainID", chainID)
+		return receipts
+	}
+
+	// check that the block number being examined is within the range of the reference data
+	if blockNumber < refs.First || blockNumber > refs.Last {
+		log.Trace("Receipt Correction: Block is out of range for receipt reference", "blockNumber", blockNumber, "start", refs.First, "end", refs.Last)
+		return receipts
+	}
+
+	// get the block nonces
+	blockNonces, ok := refs.Results[blockNumber]
+	if !ok {
+		log.Trace("Receipt Correction: Block does not contain user deposits", "blockNumber", blockNumber)
+		return receipts
+	}
+
+	// iterate through the receipts and transactions to correct the deposit nonce
+	// user deposits should always be at the front of the block, but we will check all transactions to be sure
+	udCount := 0
+	for i := 0; i < len(receipts); i++ {
+		r := receipts[i]
+		tx := transactions[i]
+		from, err := types.Sender(types.LatestSignerForChainID(tx.ChainId()), tx)
+		if err != nil {
+			log.Warn("Receipt Correction: Failed to determine sender", "err", err)
+			continue
+		}
+		// user deposits are type 126 and are not sent by the system address
+		if r.Type == 126 && from != common.HexToAddress("0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001") {
```

Comment:
126 (deposit tx type) has a constant & `0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001` should get a constant

Reply:
Can do! Addresses can't be constant but the string value can be.